etl/extract.py

```diff
@@ -4,22 +4,175 @@
-        while next_chunk:
-            logging.info(f'Retrieving monitoring locations: {next_chunk}')
-            resp = session.get(next_chunk)
-            payload = resp.json()
-            results.extend(payload['results'])
-            next_chunk = payload['next']
-
-    logging.info(f'Finished retrieving {len(results)} monitoring locations.')
-    return results
+class Extract:
+    def __init__(self):
+        """Number of records to fetch in at once"""
+        self.FETCH_LIMIT = 8
+        """Number of times to retry when there is a network error"""
+        self.FETCH_TRIES_FOR_NETWORK_ERROR = 2
+        """Number of times to retry when there is an HTTP Status code other than ok"""
+        self.FETCH_TRIES_FOR_STATUS_CODE = 2
+        """Number of times to retry when there is the JSON fails to decode"""
+        self.FETCH_TRIES_FOR_JSON = 2
+        """Number of times to accept a missing data for aborting the ETL"""
+        self.FETCH_JSON_ERROR_TOLERANCE = 2
+        """Number of seconds to wait before a retry is attempted"""
+        self.FETCH_RETRY_DELAY = 10
+        """Number of fetches per info log"""
+        self.FETCHES_PER_LOG = 128
+
+    def get_monitoring_locations(self, registry_ml_endpoint):
+        """
+        Get the monitoring location data.
+        """
+        # initialize state of errors, URL, and results
+        json_fail_count = 0
+        fetches = 0
+        url = self.construct_url(registry_ml_endpoint)
+        results = []
+
+        with self.session() as session:
+            while url:
+                if fetches % self.FETCHES_PER_LOG == 0:
+                    logging.info(f'Retrieving monitoring locations: {url}')
+                try:
+                    fetches += 1
+                    payload = self.fetch_record_block(url, session)
+                    results.extend(payload.get('results'))
+                    if payload.get('next') is None or payload.get('next') == '':
+                        break
+                except JSONDecodeError as json_err:
+                    json_fail_count += 1
+                    if json_fail_count >= self.FETCH_JSON_ERROR_TOLERANCE:
+                        logging.error('Abort: JSON errors exceeded. Set FETCH_JSON_ERROR_TOLERANCE to fine tune.')
+                        raise JSONDecodeError('JSON errors exceeded.', json_err.doc, json_err.pos)
+                    else:
+                        logging.warning(
+                            f'JSON error occurred, {self.FETCH_JSON_ERROR_TOLERANCE-json_fail_count} before abort.')
+                except RequestException:
+                    logging.error(f'Unrecoverable error fetching data from {url}')
+                    return []
+                url = self.construct_url(url)
+
+        logging.info(f'Finished retrieving {len(results)} monitoring locations.')
+        return results
+
+    # noinspection PyMethodMayBeStatic
+    # pylint: disable=too-few-public-methods, no-member
```

Comment:
I thought I would take a peak at this. My example in Teams was that just an example. This disables other issues, not the static method problem, so should probably be removed.

Reply:
That is interesting because when I search on that pylint it was in a post containing both. I will look up other pylint to see what is more relevant.

Reply:
I found the proper directive for no-self-use.